router/internal/httpclient/traceclient.go

```diff
@@ -0,0 +1,147 @@
+package httpclient
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"net/http/httptrace"
+	"sync"
+	"time"
+)
+
+type ClientTraceContextKey struct{}
+
+type TraceInjectingRoundTripper struct {
+	base http.RoundTripper
+}
+
+func NewTraceInjectingRoundTripper(base http.RoundTripper) *TraceInjectingRoundTripper {
+	return &TraceInjectingRoundTripper{
+		base: base,
+	}
+}
+
+func GetClientTraceFromContext(ctx context.Context) *ClientTraceInfo {
+	value := ctx.Value(ClientTraceContextKey{})
+	// Return no-op context if the subgraph context key was never set
+	if value == nil {
+		return &ClientTraceInfo{}
+	}
+	return value.(*ClientTraceInfo)
+}
+
+func (t *TraceInjectingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	trace := t.getClientTrace(req.Context())
+
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), trace))
+	trip, err := t.base.RoundTrip(req)
+
+	return trip, err
+}
+
+func IncrementRetryCount(ctx context.Context) {
+	fromContext := GetClientTraceFromContext(ctx)
+	fromContext.RetryCountLatestIndex++
+	fromContext.ClientTraces = append(fromContext.ClientTraces, &ClientTrace{})
```

Comment:
Why do we need to collect them?

Reply:
We right now record all the information on the retries into the metric store because there could be information (latency, etc) spikes there that might be useful for the user, esp if its a common thing.

Reply:
Why do we have special handling for retries? We are really only interested in durations. A retry would emit another data point. We don't need to distinguish them yet.